File: app/src/main/java/me/echeung/moemoekyun/api/cache/SongsCache.java
package me.echeung.moemoekyun.api.cache;

import android.util.Log;

import java.util.GregorianCalendar;
import java.util.List;

import me.echeung.moemoekyun.api.callbacks.SongsCallback;
import me.echeung.moemoekyun.api.clients.APIClient;
import me.echeung.moemoekyun.model.SongListItem;

/**
 * A naive cache of the songs data from the API for faster loading/searching.
 */
public class SongsCache {

    private static String TAG = SongsCache.class.getSimpleName();

    private static final int MAX_AGE = 1000 * 60 * 60 * 24;  // 24 hours

    private APIClient apiClient;

    private List<SongListItem> cachedSongs;
    private long lastUpdated = 0L;

    public SongsCache(APIClient apiClient) {
        this.apiClient = apiClient;

        // Prime the cache
        getSongs(null);
    }

    public void getSongs(Callback callback) {
        if (lastUpdated != 0L && isCacheValid() && cachedSongs != null && callback != null) {
            callback.onRetrieve(cachedSongs);
        }

        apiClient.getSongs(new SongsCallback() {
            @Override
            public void onSuccess(List<SongListItem> songs) {
                lastUpdated = new GregorianCalendar().getTimeInMillis();
                cachedSongs = songs;

                if (callback != null) {
                    callback.onRetrieve(cachedSongs);
                }
            }

            @Override
            public void onFailure(String message) {
                Log.e(TAG, message);
                if (callback != null) {
                    callback.onFailure(message);
                }
            }
        });
    }

    private boolean isCacheValid() {
        return new GregorianCalendar().getTimeInMillis() - lastUpdated < MAX_AGE;
    }

    public interface Callback {
        void onRetrieve(List<SongListItem> songs);
        void onFailure(String message);
    }

}

File: app/src/main/java/me/echeung/moemoekyun/client/api/cache/SongsCache.kt
package me.echeung.moemoekyun.client.api.cache

import android.util.Log
import me.echeung.moemoekyun.client.api.APIClient
import me.echeung.moemoekyun.client.api.callback.SongsCallback
import me.echeung.moemoekyun.client.model.SongListItem
import java.util.GregorianCalendar

/**
 * A naive cache of the songs data from the API for faster loading/searching.
 */
class SongsCache(private val apiClient: APIClient) {

    private var cachedSongs: List<SongListItem>? = null
    private var lastUpdated = 0L

    private val isCacheValid: Boolean
        get() = GregorianCalendar().timeInMillis - lastUpdated < MAX_AGE

    init {
        // Prime the cache
        getSongs(null)
    }

    fun getSongs(callback: Callback?) {
        if (lastUpdated != 0L && isCacheValid && cachedSongs != null && callback != null) {
            callback.onRetrieve(cachedSongs)
        }

        apiClient.getSongs(object : SongsCallback {
            override fun onSuccess(songs: List<SongListItem>) {
                lastUpdated = GregorianCalendar().timeInMillis
                cachedSongs = songs

                callback?.onRetrieve(cachedSongs)
            }

            override fun onFailure(message: String?) {
                Log.e(TAG, message)
                callback?.onFailure(message)
            }
        })
    }

    interface Callback {
        fun onRetrieve(songs: List<SongListItem>?)

        fun onFailure(message: String?)
    }

    companion object {
        private val TAG = SongsCache::class.java.simpleName

        private const val MAX_AGE = 1000 * 60 * 60 * 24 // 24 hours
    }
}
